linker/validate.go

```diff
@@ -232,7 +317,82 @@ func (r *result) validatePacked(fd *fldDescriptor, handler *reporter.Handler) er
+		err := handler.HandleErrorf(info, "packed option is only allowed on numeric, boolean, and enum fields")
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *result) validateFieldFeatures(fld *fldDescriptor, handler *reporter.Handler) error {
+	if msg, ok := fld.Parent().(*msgDescriptor); ok && msg.proto.GetOptions().GetMapEntry() {
+		// Skip validating features on fields of synthetic map entry messages.
+		// We blindly propagate them from the map field's features, but some may
+		// really only apply to the map field and not to a key or value entry field.
+		return nil
+	}
+	features := fld.proto.GetOptions().GetFeatures()
+	if features == nil {
+		// No features to validate.
+		return nil
+	}
+	if features.FieldPresence != nil {
+		switch {
+		case fld.proto.OneofIndex != nil:
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetFieldPresenceTag)
+			if err := handler.HandleErrorf(span, "oneof fields may not specify field presence"); err != nil {
+				return err
+			}
+		case fld.Cardinality() == protoreflect.Repeated:
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetFieldPresenceTag)
+			if err := handler.HandleErrorf(span, "repeated fields may not specify field presence"); err != nil {
+				return err
+			}
+		case fld.IsExtension():
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetFieldPresenceTag)
+			if err := handler.HandleErrorf(span, "extension fields may not specify field presence"); err != nil {
+				return err
+			}
+		case fld.Message() != nil && features.GetFieldPresence() == descriptorpb.FeatureSet_IMPLICIT:
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetFieldPresenceTag)
+			if err := handler.HandleErrorf(span, "message fields may not specify implicit presence"); err != nil {
+				return err
+			}
+		}
+	}
+	if features.RepeatedFieldEncoding != nil {
+		if fld.Cardinality() != protoreflect.Repeated {
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetRepeatedFieldEncodingTag)
+			if err := handler.HandleErrorf(span, "only repeated fields may specify repeated field encoding"); err != nil {
+				return err
+			}
+		} else if !internal.CanPack(fld.Kind()) && features.GetRepeatedFieldEncoding() == descriptorpb.FeatureSet_PACKED {
+			span := r.findOptionSpan(fld, internal.FieldOptionsFeaturesTag, internal.FeatureSetRepeatedFieldEncodingTag)
+			if err := handler.HandleErrorf(span, "only repeated primitive fields may specify packed encoding"); err != nil {
+				return err
+			}
+		}
+	}
+	if features.Utf8Validation != nil {
+		isMap := fld.IsMap()
+		if (!isMap && fld.Kind() != protoreflect.StringKind) ||
+			(isMap &&
+				fld.MapKey().Kind() != protoreflect.StringKind &&
+				fld.MapValue().Kind() != protoreflect.StringKind) {
```

Comment:
OOI the map field validation is ignored for protoc, this is an improvement?

Reply:
I'm not following.

If you mean the extra checks that protoc does, we don't do those checks because the parser doesn't allow producing a descriptor that is invalid per those checks.

If you mean something else related to what protoc is doing as it pertains to these highlighted lines, could you elaborate/clarify?

Reply:
Was referring to this comment: https://github.com/protocolbuffers/protobuf/blob/d4525cb9edb196e3c52e75c0e4e88f14891b0c6f/src/google/protobuf/descriptor.cc#L7902-L7906
But I misunderstood and missed the map type check here: https://github.com/protocolbuffers/protobuf/blob/d4525cb9edb196e3c52e75c0e4e88f14891b0c6f/src/google/protobuf/descriptor.cc#L7938
It looks the same. Thanks.

Reply:
Okay. As far as the former -- ignoring the checks for maps -- we do the same:
https://github.com/bufbuild/protocompile/pull/264/files#diff-e5d9143347aae9ea74aa78552eec4f014a560525d790d31f3ae317fd188c6fdbR380-R385